### PR TITLE
ADDING SMAC3 to Sparkle

### PR DESCRIPTION
**Pull request by thijssnelleman**
-  Created SMAC3 class to handle configuration calls
-  Created SMAC3 scenario class to wrap around smac.scenario.Scenario class
- Created TargetRunner subclass for Sparkle to make sure solver metrics are done through RunSolver and not Python.time lib or Pynisher.
-  Added all scenario parameters to be able to set through Sparkle
-  Added one non-scenario parameter (Initial design parameter) max_ratio
-  Set smac facade to be user-configurable
-  Added several unittests
-  Verified configure solver for runtime example works (Improved about one order size compared to SMAC2 with same budget)
-   Still lacks platform support to make use of SMAC3 MO capabilities
- SMAC3 lacks support to differentiate between Wall and CPU time

### CODE REVIEW

File: sparkle/platform/settings_objects.py [hunk 1]
```diff
@@ -906,6 +1035,232 @@ def get_smac2_max_iterations(self: Settings) -> int | None:
         max_iterations = self.__settings["smac2"]["max_iterations"]
         return int(max_iterations) if max_iterations.isdigit() else None
 
+    # Configuration: SMAC3 specific settings ###
```
Comment (jeroenrook) [inline sparkle/platform/settings_objects.py:1038]: This approach looks not very maintainable for future integrations of other AAC/HPO frameworks into Sparkle. Make ticket to make this process easier/better maintainable.

**Review by jeroenrook — CHANGES_REQUESTED** [review]

File: sparkle/types/__init__.py [hunk 2]
```diff
@@ -23,7 +24,7 @@ def _check_class(candidate: Callable) -> bool:
 def resolve_objective(objective_name: str) -> SparkleObjective:
     """Try to resolve the objective class by (case-sensitive) name.
 
-    convention: objective_name(variable-k)?(:[min|max])?
+    convention: objective_name(variable-k)?(:[min|max])?(:metric)?
```
[the remark this answers is not in the document]
Comment (thijssnelleman) [reply sparkle/types/__init__.py:27]: Not one hundred percent sure what is meant here, but updating docstring

File: sparkle/configurator/implementations/smac3.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+            name=self.name,
+            output_directory=self.results_directory / smac3_output_directory,
+            deterministic=solver.deterministic,
+            objectives=[o.name for o in self.sparkle_objectives],
+            crash_cost=crash_cost,
+            termination_cost_threshold=termination_cost_threshold,
+            walltime_limit=walltime_limit,
+            cputime_limit=cputime_limit,
+            n_trials=solver_calls or int(
+                self.cutoff_time / cputime_limit or walltime_limit),
```
[the remark this answers is not in the document]
Comment (thijssnelleman) [reply sparkle/configurator/implementations/smac3.py:333]: If we don't specify n_trials at all, SMAC3 will set it to some arbitrary value. This is a 'logical estimate' based on the users other inputs, but was not specified clearly, has been updated as:
If cutoff time is given:

1. If cpu time is specified, calculate how many runs can fit in it
2. Elif same for wall
3. Else, default to SMAC3 Default

File: sparkle/platform/settings_objects.py (hunk 1 above)
Comment (thijssnelleman) [reply sparkle/platform/settings_objects.py:1038]: Agreed, but we should I think in general discuss an overhaul of the settings class as it is too much code to maintain a single setting

**Pull request merged**